juju/juju
# [JUJU-4257] - Grant secret

This PR implements `grant-secret` and `revoke-secret` commands for user-supplied secrets.

## Checklist


## QA steps

```
cat ./testcharms/charm-repo/kubernetes/ubuntu/config.yaml
options:
  foo:
    description: a secret
    type: secret

cat ./testcharms/charm-repo/kubernetes/ubuntu/hooks/config-changed
#!/bin/bash
set -eux

foo="$(config-get foo)"
juju-log -l INFO "config-get foo=$foo"
content=$(secret-get $foo)
juju-log -l INFO "secret-get $foo=$content"

juju deploy ./testcharms/charm-repo/kubernetes/ubuntu --resource ubuntu_image=ubuntu
Located local charm "ubuntu-k8s", revision 2
WARNING deploying podspec charm "ubuntu-k8s": podspec charms are deprecated. Support for them will be removed soon.
Deploying "ubuntu-k8s" from local charm "ubuntu-k8s", revision 2 on ubuntu@20.04/stable

juju add-secret token=34ae35facd4  --label label --info "this is a secret"
secret:cj5k7v678ohs79t2aa30

juju config ubuntu-k8s foo=a
ERROR parsing settings for application: secret URI "a" not valid

juju config ubuntu-k8s foo=cj5k7v678ohs79t2aa30
ERROR updating charm config settings: option "foo" expected secret, got "cj5k7v678ohs79t2aa30"

juju config ubuntu-k8s foo=secret:cj5k7v678ohs79t2aa30

juju config ubuntu-k8s foo
secret:cj5k7v678ohs79t2aa30

juju debug-log --color --replay -m t1
application-ubuntu-k8s: 16:27:56 WARNING unit.ubuntu-k8s/2.config-changed ++ config-get foo
application-ubuntu-k8s: 16:27:56 WARNING unit.ubuntu-k8s/2.config-changed + foo=secret:cj5k7v678ohs79t2aa30
application-ubuntu-k8s: 16:27:56 WARNING unit.ubuntu-k8s/2.config-changed + juju-log -l INFO 'config-get foo=secret:cj5k7v678ohs79t2aa30'
application-ubuntu-k8s: 16:27:56 INFO unit.ubuntu-k8s/2.juju-log config-get foo=secret:cj5k7v678ohs79t2aa30
application-ubuntu-k8s: 16:27:56 WARNING unit.ubuntu-k8s/2.config-changed ++ secret-get secret:cj5k7v678ohs79t2aa30
application-ubuntu-k8s: 16:27:56 WARNING unit.ubuntu-k8s/2.config-changed ERROR permission denied
application-ubuntu-k8s: 16:27:56 WARNING unit.ubuntu-k8s/2.config-changed + content=

juju grant-secret cj5k7v678ohs79t2aa30 ubuntu-k8s

juju debug-log --color --replay -m t1
application-ubuntu-k8s: 16:28:38 WARNING unit.ubuntu-k8s/2.config-changed ++ config-get foo
application-ubuntu-k8s: 16:28:38 WARNING unit.ubuntu-k8s/2.config-changed + foo=secret:cj5k7v678ohs79t2aa30
application-ubuntu-k8s: 16:28:38 WARNING unit.ubuntu-k8s/2.config-changed + juju-log -l INFO 'config-get foo=secret:cj5k7v678ohs79t2aa30'
application-ubuntu-k8s: 16:28:38 INFO unit.ubuntu-k8s/2.juju-log config-get foo=secret:cj5k7v678ohs79t2aa30
application-ubuntu-k8s: 16:28:38 WARNING unit.ubuntu-k8s/2.config-changed ++ secret-get secret:cj5k7v678ohs79t2aa30
application-ubuntu-k8s: 16:28:38 WARNING unit.ubuntu-k8s/2.config-changed + content='token: 34ae35facd4'
application-ubuntu-k8s: 16:28:38 WARNING unit.ubuntu-k8s/2.config-changed + juju-log -l INFO 'secret-get secret:cj5k7v678ohs79t2aa30=token: 34ae35facd4'
application-ubuntu-k8s: 16:28:38 INFO unit.ubuntu-k8s/2.juju-log secret-get secret:cj5k7v678ohs79t2aa30=token: 34ae35facd4

juju exec --unit ubuntu-k8s/0 -- secret-get cj5k7v678ohs79t2aa30 --label label
token: 34ae35facd4

juju exec --unit ubuntu-k8s/0 -- secret-get --label label
token: 34ae35facd4
```

## Documentation changes

https://discourse.charmhub.io/t/command-grant-secret/11290
https://discourse.charmhub.io/t/command-revoke-secret/11291

## Bug reference

No
